zebra-network/src/peer_set/initialize.rs

```diff
@@ -918,7 +918,7 @@ where
                             "crawling for more peers in response to the crawl timer"
                         );
 
-                        crawl(candidates, demand_tx).await?;
+                        crawl(candidates, demand_tx, true).await?;
```

Comment:
I think this solution would fix this specific bug, but it would also cause Zebra to reach its maximum number of outbound connections after 1-2 hours. (The limit is 75 and there is a crawl every minute.)

At the moment, Zebra only makes new connections when it gets new peers or doesn't have enough peers for its requests. So that limits its outbound connections.

Instead, we could always add demand when `active_outbound_connections.update_count()` is below some fixed number of peers. The existing initial peer connection limit seems like a good number, or maybe half of that (rounded up).

Reply:
It checks that it's less than `peerset_outbound_connection_limit()` here: https://github.com/ZcashFoundation/zebra/blob/fix-reconnect-bug/zebra-network/src/peer_set/initialize.rs#L838

> Zebra only makes new connections when it gets new peers or doesn't have enough peers for its requests.

It was already demanding `MorePeers` whenever the crawl finds some peer addresses, I was hoping not to change its typical behaviour.

Reply:
Updated in https://github.com/ZcashFoundation/zebra/pull/7853/commits/5f3bf474307b9935ad409b536719690f2d9062fe.

Reply:
I wonder if it updates the connection tracker correctly.

Reply:
That requires the peer set to drop failed connections, I'll open a PR for that soon.